src/main/scala/com/rbmhtechnology/calliope/scaladsl/TransactionalEventProducer.scala

```diff
@@ -203,68 +253,140 @@ private class GraphRunnerSupervisor(graph: ProducerGraph, onFailure: FailureHand
 
   import ProducerGraphRunner._
   import context.dispatcher
 
   implicit val materializer = ActorMaterializer()
 
-  private def runGraph(): SourceQueueWithComplete[Unit] = {
-    val (queue, ft) = graph.run()
+  private def runGraph(): (SourceQueueWithComplete[Unit], UniqueKillSwitch) = {
+    val ((queue, ks), ft) = graph.run()
 
-    ft onComplete { result =>
-      val cause = result match {
-        case Failure(err) => err
-        case Success(_) => new UnexpectedStreamCompletionException("Transactional producer stream stopped unexpectedly")
-      }
-      self ! RestartGraph(cause)
+    ft recover {
+      case _: GraphAbortedException => Done
+    } onComplete {
+      case Failure(err) => self ! RestartGraph(err)
+      case Success(_) => self ! GraphStopped
     }
-    queue
+    (queue, ks)
   }
 
-  private var commitQueue: SourceQueueWithComplete[Unit] = _
+  startWith(Stopped, Data())
+
+  when(Stopped) {
+    case Event(RunGraph, data) =>
+      log.info("Starting transactional producer stream.")
+      val (queue, ks) = runGraph()
+      goto(Running) using data.withQueue(queue).withKillSwitch(ks) replying GraphStarted
 
-  override def preStart(): Unit = {
-    self ! StartGraph
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
   }
 
-  override def postStop(): Unit = {
-    materializer.shutdown()
-    super.postStop()
+  when(Running) {
+    case Event(RunGraph, _) =>
+      stay replying GraphStarted
+
+    case Event(StopGraph, data) =>
+      log.info("Shutting down transactional producer stream.")
+      data.killSwitch.foreach(_.abort(GraphAbortedException()))
+      goto(Stopping) using data.withNotificationFor(sender())
+
+    case Event(RestartGraph(failure), data) =>
+      log.warning("Transactional producer stream was shutdown with cause: {}. Producer will be restarted", failure)
+      val (queue, ks) = runGraph()
+      stay using data.withQueue(queue).withKillSwitch(ks)
+
+    case Event(NotifyCommit, data) =>
+      data.commitQueue.foreach(_.offer(Unit))
+      stay
   }
 
-  private def initializing: Receive = {
-    case StartGraph =>
-      log.info("Starting transactional producer.")
-      commitQueue = runGraph()
+  when(Stopping) {
+    case Event(RunGraph, _) =>
+      stash()
+      stay
+
+    case Event(StopGraph, _) =>
+      stay replying GraphStopped
+
+    case Event(GraphStopped, data) =>
+      log.info("Transactional producer stream stopped.")
+      data.notifyOnStop.foreach(_ ! GraphStopped)
       unstashAll()
-      context.become(initialized)
+      goto(Stopped) using data.withoutNotifications()
+  }
 
-    case RestartGraph(_) =>
-      log.info("Restart of transactional producer requested while initialization in progress. Restart request is ignored.")
+  whenUnhandled {
+    case Event(NotifyCommit, _) =>
+      stay
 
-    case NotifyCommit =>
-      stash()
+    case Event(ev, _) =>
+      log.info(s"Received event $ev in state $stateName. Event is ignored.")
```

Comment:
FSM was chosen because I would argue that it reads better with multiple states, the current state can easily be queried in `preStart()` (otherwise the state would have to be tracked manually) and the approach to handle the state-data is arguably better, passing around new state instead of directly changing mutable variables of the actor.

Regarding the default case in FSM - we would also have to specify the defaults in all behaviours if context-switches were used to prevent unhandled messages from being sent on the event stream so I don't see a definitive advantage of context-switches here.

I agree that the info-logging is not strictly needed in the `Unhandled` case so we could simplify it to something like this
```
  whenUnhandled {
    case Event(ev, _) =>
      stay
  }
```

WDYT?